### ISSUE TITLE
Map Authors Cannot Set an Override For Themselves

### ISSUE THREAD

**Issue by Enverex** [Frooxius/NeosPublic#768]
I ran into an issue due to a combination of circumstances and it's made it a real issue for hosting worlds on a headless.

When creating a world, if someone joins, you can manually set permissions for that person and they'll be saved into the world. That works fine. The issue arises given that the map author has no override set for themselves by default, it's just "assumed". They can't change their own permission level (which would normally create the override) as they are always locked at Admin. This means that if someone else hosts a session of that world, all the overrides that have been set will work, but if the author themselves joins, they'll be assigned a default role position instead.

This becomes a deadlocked scenario on headless where if you're hosting someone's world for them, they will have have no permissions because they're treated like a basic "Registered User" rather than the author and there's no way of promoting them.

For those reasons, I'd request that by default a hard "admin" override is set for the author of a world. Obviously anything cleaner is welcomed - but as it stands, it means people can't maintain / moderate their own worlds if they aren't also the host.

**Comment by shiftyscales** [comment]
As discussed on Discord, I believe a potentially cleaner solution would involve having a UI to manage world overrides for world creators.

It could allow users to add/remove persistent overrides for any arbitrary users without said users being present in the world.

There are some considerations that could be made, e.g. not enforcing the overrides unless explicitly requested, e.g. So session hosts don't grant unintended permissions to the world creator/their friends on any arbitrary session of the world that is hosted.

This system could perhaps be tied to the planned user tagging system, perhaps, so permission overrides could be added, e.g. for users of a particular tag, or group.

It would be a quality of life, and functional change to have a visual way to manage overrides. Right now the only way to clear them is to remove them all, then save the world. Additionally, unintended overrides can be made at present if the host grants permissions, then forgets to clear them before saving.

Perhaps it would be best that overrides are only made manually?

Another factor to consider: world access- this too could be tied in to the user tagging/friend system?

**Comment by Enverex** [comment]
> As discussed on Discord, I believe a potentially cleaner solution would involve having a UI to manage world overrides for world creators.

The reason I didn't mention that for now is that that would likely require a decent amount of work, pushing it back quite some way. As it stands, the headless is virtually useless to me as I can't admin/moderate my own worlds, it also means I can't host maps for others either due to them not having permissions on their own worlds. Unfortunately the other workarounds mentioned earlier are either very messy or simply won't work given other circumstances.

**Comment by Frooxius** [comment]
Overriding the author as admin won't work, because it would violate the principles of the permission system and open lots of doors for abuse. The system assumes that the host always has the highest permissions, but with this you could end up with scenario where you simply joining someone in a map you created get more privileges over the session than the person actually hosting it.

E.g. in the future someone could publish a map and if it gets widely used, they could mess with people's sessions, simply by joining them.

If you want to have Admin on your headless, it needs to be done the other way - the account hosting the world needs to have Admin role. If it has Admin role, then it can grant you the same role. The system can't ever grant a role higher than it has itself.

I could add a code that takes per-user override into account. You as map creator would assign Admin role to your headless account and save the world.

**Comment by Enverex** [comment]
That's making a few assumptions though. For example if someone wants me to host a world for them (on a headless), they wouldn't necessarily want me (and by extension my headless) to have any higher rights than a normal user. They would, however, still want the saved overrides to apply (and thus keep admin themselves).

Regarding my own maps, I can't add an override for the headless as it's not in the world and you can only set overrides for people who are present.

>The system assumes that the host always has the highest permissions, but with this you could end up with scenario where you simply joining someone in a map you created get more privileges over the session than the person actually hosting it.

That doesn't seem completely unexpected given that would only ever apply to one person (the map creator).

**Comment by Frooxius** [comment]
That's the whole point of the permission system though. If they are not willing to give you (or any host) higher permissions, you won't get them. The owner of the world decides what permissions others get.

For adding overrides for accounts that are not present, that can be added, it's part of the feature proposal.

**Comment by Enverex** [comment]
That still leaves the primary issue that a map author can add overrides for other people, but not themselves.

**Comment by shiftyscales** [comment]
Hence his statement "For adding overrides for accounts that are not present, that can be added, it's part of the feature proposal."

That mechanism would allow adding overrides for any user, presumably also including the map author.

**Comment by Frooxius** [comment]
That's not really a problem, you can add that in the Config.json for the headless. The only problem is that the host can ever grant you permission as high as it has itself. So if the headless account is Builder, it can't give you Admin. This applies to all other people too,  you can override someone as Admin on your map, but if the person who hosts it is lower than that, that override won't apply.

**Comment by Enverex** [comment]
We're kinda going around in circles here. It is a problem because the headless won't have permission so it can't override grant someone permissions. But it doesn't/shouldn't need them. Being able to set overrides for other people *but not yourself* is clearly a bug/oversight. It makes no sense to be be able to set other people's permission overrides but not your own.

Example last night: My headless was hosting someone's map. The headless was spectator due to no overrides, but when I joined I was still Builder because the author had explicitly given me that rank as an override when he was originally working on the world. Unfortunately he was also stuck as spectator as he was the author and thus had no override and got stuck at the default.

**Comment by TehTurk** [comment]
Originally this was okay before when you had the record url you yourself could set whatever you want. I always wondered why the permissions was the change vs who had access to the record url. Only the owner of the world should have access or whomever they decide to give it too.

**Comment by shiftyscales** [comment]
For security, and the protection of a world creator's creations, @TehTurk. E.g. prior to those changes, any user could host the session with their role set to admin/builder, modify the world as they pleased, e.g. making it saveable when it wasn't prior, etc. and take their own copy/assets from it.

If a URL is shareable at all, then it can't be considered secure. If a world is public, it's public.

**Comment by TehTurk** [comment]
I'm aware and agree, but it's also an issue of who should have access to that URL. Unless your world creator you shouldn't have access to the world url typically, and should only be shared in certain host sharing or other situations. The world permissions should take precedent, but if your hosting a world, typically you'll need some some of control or proper host/control.

**Comment by TehTurk** [comment]
Sharing the URL or Giving someone permissions and saving the override are no different, as it still requires trust between the users.

**Comment by Enverex** [comment]
Everyone has access to every world URL. You can access them via the World Orbs and the logs. But that's something of a tangent to my reported issue here.

**Comment by TehTurk** [comment]
Hmm nevermind me then.

**Comment by shiftyscales** [comment]
Additionally as I had noted in my initial proposal: 
> Another factor to consider: world access- this too could be tied in to the user tagging/friend system?

If that were implemented, then even though the URL itself was public, only certain users could have access to open the world, similar to how right now it's possible to have Patron-only worlds, @TehTurk. E.g. users from a particular tag, or group the world creator has specified.

It's not applicable to this specific issue, but rather just another aspect to be potentially considered when the user tagging system is implemented.

**Comment by Frooxius** [comment]
I've already proposed simply setting up an override through the Config.json. You can define ```"defaultUserRoles": { "Enverex" : "Admin" },``` there, which will create an override for you. Is there a reason why this solution isn't acceptable right now?

I've also proposed adding on UI for setting up overrides when making the map for any users, even if they're not present, or yourself.

From my understanding, those two combined would resolve the issue you're having. If not, can you elaborate a bit on why are they problematic please?

What you've originally proposed with the owner always being admin is simply not possible like I said in my initial post. It would violate the design of the permission system and open up holes for abuse. As such it's just not an option, so if we continue along that path we'll keep going in circles. The problem needs to be addressed in ways that don't violate the permission system.

**Comment by Enverex** [comment]
Because it doesn't work...
`"12:39:43.927    Role Admin not available for world: ZombieDoll Lounge`

A UI for manually setting overrides would work, but seems more like a long term fix where as I was hoping for some sanity fix in the meantime.

**Comment by Frooxius** [comment]
I think you are misunderstanding what it is that I'm exactly proposing. To clarify any misunderstandings there are a few important points:

- The reason you're getting that message is because the headless host doesn't have the Admin role itself, but a lower one
- The host cannot ever assign a role higher than it has itself. If the host is Builder, it cannot give anyone Admin. **This is non-negotiable**. No UI or default permission for owners or anything else will ever be able to violate that.
- If you want your headless to give you (or anyone else) the Admin role, you need to grant it the Admin role
- What I'm proposing is adding a mechanism so you can setup an override for accounts that are not present and save them with your world. If you don't want the "Default Host" to be Admin, you would create override for "EnverexBot" to be "Admin" so it has permission to grant this role (either through Config.json or through an override)
- The feature above could also be used to create an override for yourself saved with the world. However if the host has role lower than that, it still won't be able to grant you that role.

In short, this issue cannot be addressed without ensuring the host account has the Admin role first. That's the main problem there. 

I can also tweak the role system on headless so instead of rejecting the role, it just uses the highest available instead. That should help in some other cases a bit at least.

**Comment by Enverex** [comment]
I think the confusion stems from the fact it doesn't seem to work as you've described there, unless I'm misreading.

>The host cannot ever assign a role higher than it has itself. If the host is Builder, it cannot give anyone Admin. This is non-negotiable. No UI or default permission for owners or anything else will ever be able to violate that.

When I joined the headless session (the headless itself was Spectator) I was granted Builder as that had been saved as an override when the map was created. Going by what you described, that shouldn't have been possible as the headless (the host) wasn't Builder or Admin.

**Comment by shiftyscales** [comment]
You are misreading. It's not the host assigning that role, @Enverex. It's the override you have inside of the map that is assigning that override. Overrides take precedence over the defaults that are assigned by the host, hence they are overrides.

**Comment by Frooxius** [comment]
I see. That's really strange. Are you sure that host was Spectator in that case, how exactly did you check? That sounds like something is definitely quite broken. I can investigate that and see if I can see that behavior.

**Comment by shiftyscales** [comment]
I followed up with Enverex with this on Discord/in Neos:

"Keep Original Roles" is what was causing the confusion.

It appears by default that value is false, both for a user hosted, and headless hosted session, and when that is the case, the permissions are all working perfectly as you specified.

The only exception is if 'keep original roles' is enabled/true. Then a user with an override can be higher than the headless/session host.

E.g. I had EnverexBot host a world I'd created, and set myself as an admin override on. When I joined his world with keep original roles true, I was an admin while the host was a spectator.

When I had Enverex (who I'd assigned a builder override to) host on his main account, I joined, and I was a builder, but not admin, so things are working perfectly/as intended for user-hosted sessions. They just aren't working properly for headless when "Keep Original Roles" is enabled as it allows the overrides to be higher the rights of the host. E.g. I was an admin in EnverexBot's session where even the EnverexBot was only a spectator.

**Comment by Frooxius** [comment]
I found major cause of the weird behaviors, the headless host was never assigned a proper role, so it was still effectively Admin, but some of the checks were done against the default host role, causing a weird hybrid state.

However since the actual role was Admin it was able to grant the builder role to some users, since they were lower than its actual role. I'm not quite sure why do you thought it was a spectator though, I could not reproduce that.

I've fixed the headless not having a role in 2020.7.15.1186 and also made all checks against its actual role, so any overrides now be used properly.

Also the headless will now just use the highest available role, rather than not assigning it at all if the host has lower role than  you want to assign. E.g. if the host is Builder and your override is Admin, it'll still give you Builder at very least.

**Comment by shiftyscales** [comment]
To clarify- in the setup, I had created a world on my account, added an admin override for myself, a builder override for Enverex, and the default roles for each category was set to spectator.

So defaults for anon, visitors, friends, and hosts were all set to spectator.

While EnverexBot was hosting, and I had joined the session, it reported that they were a spectator on the session list (this is probably the weird hybrid state you had mentioned.) I didn't test the host actually was a spectator- but it reported that they were.
